File: iosApp/iosApp/KuiklyRenderExpand/Views/KRMyDemoCustomView.m
/*
 * Tencent is pleased to support the open source community by making KuiklyUI
 * available.
 * Copyright (C) 2025 Tencent. All rights reserved.
 * Licensed under the License of KuiklyUI;
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * https://github.com/Tencent-TDS/KuiklyUI/blob/main/LICENSE
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#import "KRMyDemoCustomView.h"
#import "KRComponentDefine.h"

@interface KRMyDemoCustomView ()
/** message 文字标签 */
@property (nonatomic, strong) UILabel *messageLabel;
/** Tap Me 按钮 */
@property (nonatomic, strong) UIButton *tapButton;
/** kuikly 属性：显示文字 */
@property (nonatomic, copy, nullable) NSString *KUIKLY_PROP(message);
/** kuikly 事件：Tap Me 点击回调 */
@property (nonatomic, strong, nullable) KuiklyRenderCallback KUIKLY_PROP(onMyViewTapped);
@end

@implementation KRMyDemoCustomView

- (instancetype)initWithFrame:(CGRect)frame {
    if (self = [super initWithFrame:frame]) {
        self.backgroundColor = [UIColor yellowColor];
        self.layer.borderColor = [UIColor blackColor].CGColor;
        self.layer.borderWidth = 2.0;

        _messageLabel = [[UILabel alloc] init];
        _messageLabel.font = [UIFont systemFontOfSize:24];
        _messageLabel.textColor = [UIColor blackColor];
        _messageLabel.textAlignment = NSTextAlignmentCenter;
        [self addSubview:_messageLabel];

        _tapButton = [UIButton buttonWithType:UIButtonTypeSystem];
        [_tapButton setTitle:@"Tap Me" forState:UIControlStateNormal];
        [_tapButton.titleLabel setFont:[UIFont systemFontOfSize:18]];
        [_tapButton addTarget:self action:@selector(onTapButtonClicked)
             forControlEvents:UIControlEventTouchUpInside];
        [self addSubview:_tapButton];
    }
    return self;
}

- (void)layoutSubviews {
    [super layoutSubviews];
    CGFloat halfH = self.bounds.size.height / 2.0;
    self.messageLabel.frame = CGRectMake(0, 0, self.bounds.size.width, halfH);
    self.tapButton.frame = CGRectMake(0, halfH, self.bounds.size.width, halfH);
}

#pragma mark - KuiklyRenderViewExportProtocol

- (void)hrv_setPropWithKey:(NSString *)propKey propValue:(id)propValue {
    // 让通用属性（含 accessibility/accessibilityRole/accessibilityInfo）与父类 css_touch* 等先生效。
    KUIKLY_SET_CSS_COMMON_PROP;
    // 自定义 css_message / css_onMyViewTapped 通过 KVC 派发到本类 setter。
}

#pragma mark - CSS Property Setters

- (void)setCss_message:(NSString *)css_message {
    _css_message = [css_message copy];
    self.messageLabel.text = _css_message ?: @"";
}

- (void)setCss_onMyViewTapped:(KuiklyRenderCallback)css_onMyViewTapped {
    _css_onMyViewTapped = css_onMyViewTapped;
}

#pragma mark - Actions

- (void)onTapButtonClicked {
    if (self.css_onMyViewTapped) {
        self.css_onMyViewTapped(@{});
    }
}

#pragma mark - Subviews Ordering

/**
 * kuikly 侧通过 addSubview: 添加的子节点默认追加在最上层；
 * 装饰视图（messageLabel / tapButton）应位于最底层，避免遮挡 kuikly 子节点。
 * 这里在每次插入 kuikly 子节点后，主动把装饰视图沉到最底部。
 */
- (void)didAddSubview:(UIView *)subview {
    [super didAddSubview:subview];
    if (subview != self.messageLabel && subview != self.tapButton) {
        [self sendSubviewToBack:self.tapButton];
        [self sendSubviewToBack:self.messageLabel];
    }
}

@end

File: androidApp/src/main/java/com/tencent/kuikly/android/demo/KRMyDemoCustomView.kt
/*
 * Tencent is pleased to support the open source community by making KuiklyUI
 * available.
 * Copyright (C) 2025 Tencent. All rights reserved.
 * Licensed under the License of KuiklyUI;
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * https://github.com/Tencent-TDS/KuiklyUI/blob/main/LICENSE
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.tencent.kuikly.android.demo

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.util.TypedValue
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.LinearLayout
import android.widget.TextView
import com.tencent.kuikly.core.render.android.expand.component.KRView
import com.tencent.kuikly.core.render.android.export.KuiklyRenderCallback

/**
 * Demo 自定义 View，对齐鸿蒙侧 `KRMyDemoCustomView.ets`：
 * - 支持 `message` 属性：居中显示文字（黄底黑边）
 * - 支持 `onMyViewTapped` 事件：Tap Me 按钮点击回调
 * - 支持 kuikly 侧嵌入的子节点（作为 FrameLayout 由框架 addChild 到本 View 上）
 *
 * 无障碍属性 `accessibility` / `accessibilityRole` / `accessibilityInfo` 走 `KRView` 基类
 * 的通用属性通道（见 `KRCSSViewExtension.kt`），此处无需重写。
 * `accessibilityAnnounce` / `accessibilityFocus` method 走 `IKuiklyRenderViewExport.call`
 * 基类实现，此处无需重写。
 */
class KRMyDemoCustomView(context: Context) : KRView(context) {

    private val decorLayout: LinearLayout
    private val messageText: TextView
    private val tapButton: Button

    private val borderPaint: Paint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
        style = Paint.Style.STROKE
        color = Color.BLACK
        strokeWidth = 2f * context.resources.displayMetrics.density
    }

    private var onMyViewTapped: KuiklyRenderCallback? = null

    init {
        setBackgroundColor(Color.YELLOW)
        setWillNotDraw(false)

        decorLayout = LinearLayout(context).apply {
            orientation = LinearLayout.VERTICAL
            gravity = Gravity.CENTER
            layoutParams = LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.MATCH_PARENT
            )
        }
        messageText = TextView(context).apply {
            setTextSize(TypedValue.COMPLEX_UNIT_SP, 24f)
            setTextColor(Color.BLACK)
            gravity = Gravity.CENTER
            layoutParams = LinearLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                0,
                1f
            )
        }
        tapButton = Button(context).apply {
            text = "Tap Me"
            layoutParams = LinearLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                0,
                1f
            )
            setOnClickListener {
                onMyViewTapped?.invoke("{}")
            }
        }
        decorLayout.addView(messageText)
        decorLayout.addView(tapButton)
        // 装饰层放在最下面，kuikly 侧后续 addChild 的子节点会盖在上面（对齐鸿蒙侧 Stack 语义）。
        addView(decorLayout)
    }

    override fun setProp(propKey: String, propValue: Any): Boolean {
        return when (propKey) {
            PROP_MESSAGE -> {
                messageText.text = propValue as? String ?: ""
                true
            }
            PROP_ON_MY_VIEW_TAPPED -> {
                onMyViewTapped = propValue as? KuiklyRenderCallback
                true
            }
            else -> super.setProp(propKey, propValue)
        }
    }

    override fun dispatchDraw(canvas: Canvas) {
        super.dispatchDraw(canvas)
        // 绘制黑色边框
        val inset = borderPaint.strokeWidth / 2f
        canvas.drawRect(inset, inset, width - inset, height - inset, borderPaint)
    }

    companion object {
        const val VIEW_NAME = "KRMyDemoCustomView"
        private const val PROP_MESSAGE = "message"
        private const val PROP_ON_MY_VIEW_TAPPED = "onMyViewTapped"
    }
}
